feat: stable Device::id()

xephyris opened the pull request:
This PR adds a `Device::id()` command that returns the OS native device id on supported operating systems.
~~Currently, only macOS and Windows (WASAPI) are supported.~~

### Function

By returning a `Device::id()`, it will be easier to ensure that the same device gets connected on application startup. This also allows for greater extension capability for other libraries to extend upon `cpal` function.

### Architecture

The OS specific device ids are all stored in a `DeviceId` enum, that allows for support for the various data types that different operating systems and APIs use to store device ids (e.g. `u32` for macOS and `String` for WASAPI.

``` rust
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum DeviceId {
    CoreAudio(String),
    WASAPI(String),
    // ..
}
```

To handle any errors that might be outputted in when running `Device::id()`, there is also a new `DeviceIdError` enum
```rust
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum DeviceIdError {
    /// See the [`BackendSpecificError`] docs for more information about this error variant.
    BackendSpecific { err: BackendSpecificError },
    UnsupportedOS,
    ParseError,
}
```

Calling `Device::id()` will return a `Result<DeviceId, DeviceIdError>`. If the OS is supported and the device id is obtained, it will return a `DeviceId`. If the OS is unsupported, a `DeviceIdError::UnsupportedOS` error will be returned.

### Current Status
~~Currently, support for Windows WASAPI `String` Ids has been added, along with support for macOS `u32` Ids. I am working to add support for `ALSA` and maybe `pulseaudio` (once that gets merged).~~

For other APIs, I do not have the hardware to test or program for them, ~~so they currently return a `DeviceIdError::UnsupportedOS` error.~~

### Testing
`Device::id()` for macOS should return the ~~standard `u32`~~ device uid that can be used anywhere. 
For Windows, to convert the Rust `String` type to a `PWSTR`, the following code can be used
```rust
// device_id is the id returned by cpal
 let mut id = format!("{}\0", device_id).encode_utf16().collect::<Vec<u16>>(); 
 let pwstr = PWSTR(id.as_mut_ptr());
```

roderickvd commented on src/lib.rs:
```diff
@@ -219,6 +219,15 @@ where
+/// Currently only supports macOS and Windows (WASAPI)
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DeviceId {
+    MacOS(u32),
```
For macOS, it's better to use the `kAudioDevicePropertyDeviceUID` instead of the AudioDeviceID, because the device UID doesn't change on reboot.

Also better to rename to `CoreAudio` as there may be multiple hosts on the macOS platform.

roderickvd commented on src/lib.rs:
```diff
@@ -219,6 +219,15 @@ where
 
+/// The device ID of the audio device, on supported OSs
+/// Currently only supports macOS and Windows (WASAPI)
+#[derive(Clone, Debug, Eq, PartialEq)]
```
You could add a variant `AAudio(i32)` with an implementation in `src/host/aaudio/mod.rs` like:

```rust
fn id(&self) -> Result<DeviceId, DeviceIdError> {
    match &self.0 {
        None => Ok(DeviceId::aaudio(-1)), // Default device
        Some(info) => Ok(DeviceId::aaudio(info.id)),
    }
}
```

And a variant `Jack(string)` with in `src/host/jack/mod.rs`:

```rust
fn id(&self) -> Result<DeviceId, DeviceIdError> {
    Ok(DeviceId::jack(self.name.clone()))
}
```

roderickvd commented on src/host/null/mod.rs:
```diff
@@ -47,6 +47,11 @@ impl DeviceTrait for Device {
 
+    #[inline]
+    fn id(&self) -> Result<DeviceId, DeviceIdError> {
+        Err(DeviceIdError::UnsupportedOS)
```
This could simply return `Ok(DeviceId::Null)`.

roderickvd commented on src/traits.rs:
```diff
@@ -89,6 +89,9 @@ pub trait DeviceTrait {
     fn name(&self) -> Result<String, DeviceNameError>;
 
+    /// The device-id of the device. (For supported OSs only)
+    fn id(&self) -> Result<DeviceId, DeviceIdError>;
```
Then on the `HostTrait`, it would be desired to have:

```rust
fn device_by_id(&self, id: &DeviceId) -> Option<Self::Device> {
     self.devices().ok()?.find(|device| {
        device.device_id().ok().as_ref() == Some(id)
    })
}
```

(or something smarter)

roderickvd commented on src/lib.rs:
```diff
@@ -219,6 +219,15 @@ where
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DeviceId {
+    MacOS(u32),
+    Windows(String),
```
Let's stick to `Wasapi` as there could be multiple hosts on the Windows platform.

roderickvd reviewed (changes requested):
This is great! Your proposal _literally_ crossed paths as I was working on something also. No matter, all the better I could provide some review comments with actual code suggestions.

Please also add a changelog, it's worth it 😄

xephyris commented:
Hello @roderickvd !

I have implemented most of the changes requested and renamed the enum variants to the suggested names.

For `jack` and `aaudio`, I just copied the code that you provided, and because I cannot test it, there may or may not be some errors.

I have also added `alsa` support which currently just returns the same content as name, because I was unable to find any other specific identifiers that could work better.

I will work on changing the macOS CoreAudio type to `kAudioDevicePropertyDeviceUID` in the coming days.

The changelog has also been updated 😄.

roderickvd commented:
Cool stuff let me know when you're ready for me to take another look.

roderickvd commented on src/host/coreaudio/macos/device.rs:
```diff
@@ -392,6 +403,37 @@ impl Device {
+                NonNull::from(&mut name).cast(),
+            )
+        };
+        if status == 0 {
```
Instead of checking `status == 0` would it make sense to `check_os_status(status)?` consistent with the rest of the code?

Then instead (or additionally) we should return an error `if uid.is_null()`.

roderickvd commented on src/host/coreaudio/macos/device.rs:
```diff
@@ -392,6 +403,37 @@ impl Device {
+            mScope: kAudioObjectPropertyScopeGlobal,
+            mElement: kAudioObjectPropertyElementMain,
+        };
+        let mut name: CFStringRef = std::ptr::null_mut();
```
Is this the `name` or `uid`?

roderickvd commented on CHANGELOG.md:
```diff
@@ -1,5 +1,6 @@
 # Unreleased
 
+- Add `DeviceId` enum supporting most APIs that returns a system device id for the audio device, allowing it to be reselected using the `HostTrait::device_by_id()` function
```
⬇️ 

```markdown
- Add `HostTrait::id` method that returns a stable audio device ID.
- Add `HostTrait::device_by_id` to select a device by its stable ID.
```

roderickvd commented on src/host/asio/device.rs:
```diff
@@ -54,6 +56,10 @@ impl Device {
     }
 
+    fn id(&self) -> Result<DeviceId, DeviceIdError> {
+        Err(DeviceIdError::UnsupportedPlatform)
```
Not an ASIO user but wouldn't this work:

```rust
Ok(DeviceId::ASIO(self.driver.name().to_string()))
```

roderickvd commented on src/host/coreaudio/ios/mod.rs:
```diff
@@ -74,6 +75,10 @@ impl Device {
     }
 
+    fn id(&self) -> Result<DeviceId, DeviceIdError> {
+        Err(DeviceIdError::UnsupportedPlatform)
```
I think it would be possible to get the input/output name through `AVAudioSession::shared_instance().current_route().outputs/inputs` but right now we always seem to only enumerate and always select the default device. Would it make sense to hardcode the default device here as a value?

roderickvd commented on src/host/emscripten/mod.rs:
```diff
@@ -69,6 +69,11 @@ impl Device {
 
+    #[inline]
+    fn id(&self) -> Result<DeviceId, DeviceIdError> {
+        Err(DeviceIdError::UnsupportedPlatform)
```
Same, would it make sense to always return `default`?

roderickvd commented on src/host/webaudio/mod.rs:
```diff
@@ -84,6 +85,11 @@ impl Device {
 
+    #[inline]
+    fn id(&self) -> Result<DeviceId, DeviceIdError> {
+        Err(DeviceIdError::UnsupportedPlatform)
```
Same.

roderickvd commented on src/lib.rs:
```diff
@@ -219,6 +219,51 @@ where
+                Ok(DeviceId::AAudio(id))
+            }
+            "jack" => Ok(DeviceId::Jack(data.to_string())),
+            _ => Err(DeviceIdError::UnsupportedPlatform),
```
This may require syncing.

xephyris replied:
I'm unsure about what you mean by this, would you care to elaborate?

roderickvd replied:
Sorry I meant to say that by now you’ve added more platforms that support it, so instead of matching on `_` I propose to include all variants and your code will be “in sync”.

roderickvd reviewed (changes requested):
Some ideas and one point about the macOS implementation.

roderickvd commented:
From your PR description, which I edited to show the progress made since:

> For Windows, to convert the Rust `String` type to a `PWSTR`, the following code can be used

Is that still necessary?

xephyris commented:
Thanks for the update @roderickvd ! I'll try getting the code updated soon for another review.

> Is that still necessary?

Currently I believe it is still necessary to convert it to a `PWSTR` because to get the `IMMDevice` from the `IMMDeviceEnumerator`, it requires a `PWSTR` at least from what I recall. 

This is primarily for if you are using the `Device::id()` to extend upon `cpal` functionality like I'm currently working on `cpvc` and adding the volume controls. If you are just trying to re-select a device from `cpal`, then that code to convert to `PWSTR` isn't necessary.

xephyris commented:
So right now I'm having it return the `Device::id()` as a `String` guid for Windows and if you iterate with `device_by_id(id)` to get the device, it will work fine.

The snippet about `PWSTR` that I included is mainly for if you want to access a Windows WASAPI audio device yourself using the `id` provided by `cpal`. 

That code is more about external access using the `id`s from `cpal`, not if you are only using `cpal` to access and adjust audio devices.

xephyris commented on src/lib.rs:
```diff
@@ -272,6 +272,7 @@ impl std::str::FromStr for DeviceId {
             "emscripten" => Ok(DeviceId::Emscripten(data.to_string())),
             "ios" => Ok(DeviceId::IOS(data.to_string())),
+            "null" => Ok(DeviceId::Null),
             _ => Err(DeviceIdError::UnsupportedPlatform),
```
Unfortunately, I can't find a way to remove the `_` catch-all. This is an `&str` match so it gives me a non-exhaustive error if I do try to remove the `_` catch-all.

Here is the error below 
```
non-exhaustive patterns: `&_` not covered
the matched value is of type `&str`
`&str` cannot be matched exhaustively, so a wildcard `_` is necessary rustc[Click for full compiler diagnostic]
lib.rs(275, 41): ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown: `,
            &_ => todo!()`
```

xephyris replied:
So is there anything I should do at this point? Or should I mark this as resolved?

xephyris commented on src/lib.rs:
```diff
@@ -220,6 +220,64 @@ where
 
+/// The device ID of the audio device, on supported OSs
+/// Currently only supports macOS and Windows (WASAPI)
+#[derive(Clone, Debug, Eq, PartialEq)]
```
I originally based my implementation off of `DeviceName` and `DeviceNameError`, so this slipped through without me noticing, they are now switched. Thanks for catching this!

roderickvd commented on src/host/coreaudio/macos/device.rs:
```diff
@@ -44,6 +48,9 @@ use std::time::{Duration, Instant};
 use super::property_listener::AudioObjectPropertyListener;
 use coreaudio::audio_unit::macos_helpers::get_device_name;
+
+type CFStringRef = *mut std::os::raw::c_void;
```
Could we remove this in favor of plain `CFString` and be less unsafe?

```rust
  fn id(&self) -> Result<DeviceId, DeviceIdError> {
      let property_address = AudioObjectPropertyAddress {
          mSelector: kAudioDevicePropertyDeviceUID,
          mScope: kAudioObjectPropertyScopeGlobal,
          mElement: kAudioObjectPropertyElementMain,
      };

      // CFString is retained by the audio object, use wrap_under_get_rule
      let mut uid: *const CFString = std::ptr::null();
      let data_size = size_of::<*const CFString>() as u32;

      // SAFETY: AudioObjectGetPropertyData is documented to write a CFString pointer
      // for kAudioDevicePropertyDeviceUID. We check the status code before use.
      let status = unsafe {
          AudioObjectGetPropertyData(
              self.audio_device_id,
              NonNull::from(&property_address),
              0,
              null(),
              NonNull::from(&data_size),
              NonNull::from(&mut uid).cast(),
          )
      };

      check_os_status(status)?;

      if uid.is_null() {
          return Err(DeviceIdError::BackendSpecific {
              err: BackendSpecificError {
                  description: "Device UID is null".to_string(),
              },
          });
      }

      // SAFETY: We verified uid is non-null and the status was successful
      let uid_string = unsafe { CFString::wrap_under_get_rule(uid).to_string() };
      Ok(DeviceId::CoreAudio(uid_string))
  }
```

roderickvd commented on src/host/wasapi/device.rs:
```diff
@@ -322,6 +326,18 @@ impl Device {
+    fn id(&self) -> Result<DeviceId, DeviceIdError> {
+        unsafe {
+            match self.device.GetId() {
+                Ok(pwstr) => match pwstr.to_string() {
```
Instead of silently discarding, we could:

```rust
  fn id(&self) -> Result<DeviceId, DeviceIdError> {
      // SAFETY: Calling GetId on a valid IMMDevice is safe
      unsafe {
          let pwstr = self.device.GetId().map_err(|e| DeviceIdError::BackendSpecific {
              err: e.into(),
          })?;

          let id_str = pwstr.to_string().map_err(|e| DeviceIdError::BackendSpecific {
              err: BackendSpecificError { description: format!("Failed to convert device ID to string: {}", e) },
          })?;

          Ok(DeviceId::WASAPI(id_str))
      }
  }
```

roderickvd commented on src/host/coreaudio/ios/mod.rs:
```diff
@@ -88,6 +89,10 @@ impl Device {
     }
 
+    fn id(&self) -> Result<DeviceId, DeviceIdError> {
+        Ok(DeviceId::IOS("default".to_string()))
```
"default" differs from "Default Device" in `name`. Maybe use shared const?

roderickvd replied:
Good rationale, let’s keep it like it this then. 

How do you feel this’ll be for ALSA? Spaces in device names (e.g. for USB devices) would not be impossible.

roderickvd commented on CHANGELOG.md:
```diff
@@ -1,5 +1,7 @@
 # Unreleased
 
+- Add `HostTrait::id` method that returns a stable audio device ID.
```
Should be `DeviceTrait`.

roderickvd reviewed (commented):
Thanks for sticking through 👍 this is a pretty important feature so let's get it right!

roderickvd commented:
Thanks, I’ll review later as I’m away for a few days.

roderickvd reviewed (changes requested):
Thanks for the ping and sorry for the wait. Yes lots of work done, thanks for that. Here's a few hopefully small ones.
